File: JoinCubeF/src/main/scala/com/jpro/framework/Dispatcher.scala
package com.jpro.framework

import java.util.concurrent.ArrayBlockingQueue

import org.apache.logging.log4j.scala.Logging
import org.json4s._
import org.json4s.JsonAST.JValue
import org.json4s.jackson.JsonMethods._

import scala.util._

/**
  * The core working thread
  * You should pass processor
  * of different types to Dispatcher.
  */
class Dispatcher(q: ArrayBlockingQueue[BaseBlock],
                 f: JValue => Unit,
                 g: String => Unit)
  extends Thread with Logging {

  /**
    * One single instance of shared queue
    * for many consumer to process data in it.
    */
  val sharedQueue: ArrayBlockingQueue[BaseBlock] = q

  /**
    * core functor to process data
    */
  val core: JValue => Unit = f

  /**
    * store exceptional data
    */
  val trap: String => Unit = g

  /**
    * Dispatcher loop
    */
  override def run(): Unit = {
    while (GlobalContext.working
      || sharedQueue.size() != 0) {
      Try(sharedQueue.take()) match {
        case Failure(ex) =>
          logger.error("Take data from SharedQueue - " + ex)
        case Success(data) =>
          Try(parse(data.body)) match {
            case Failure(parseEx) =>
              logger.error(parseEx)
              trap(data.body)
            case Success(json) =>
              Try(core(json)) match {
                case Failure(procEx) =>
                  logger.error(procEx)
                  trap(data.body)
                case Success(_) =>
              }
          }
      }
    }
  }
}

File: FullInspection/src/main/java/com/jpro/Dispatcher.java
package com.jpro;

import com.alibaba.fastjson.JSON;
import com.alibaba.fastjson.JSONException;
import com.alibaba.fastjson.JSONObject;
import lombok.extern.log4j.Log4j2;

import java.util.concurrent.ArrayBlockingQueue;

@Log4j2
public class Dispatcher extends Thread {

    private SeqAct processor;

    private ArrayBlockingQueue<Msg> sharedQueue;

    private String _id;

    public Dispatcher(SeqAct p, ArrayBlockingQueue<Msg> q, String id) {
        processor = p;
        sharedQueue = q;
        _id = id;
    }

    @Override
    public void run() {
        log.info("( {} ) enter Dispatcher loop", _id);
        while (Unique.working.get() || sharedQueue.size() != 0) {
            Msg msg = null;
            try {
                msg = sharedQueue.take();
            } catch (InterruptedException ie) {
                log.error("( {} ) BlockQueue take data error - {}", _id, ie);
            }
            if (msg == null) continue;

            if (msg.head.equals("exit")) {
                break;
            }

            JSONObject jo = null;
            try {
                jo = JSON.parseObject(msg.body);
            } catch (JSONException je) {
                log.error("( {} ) JSON parse error - {}", _id, je);
                continue;
            }
            if (jo == null) continue;

            try {
                if (!processor.validate(jo)) {
                    continue;
                }
            } catch (RuntimeException re) {
                log.error("( {} ) validate failed - {}", _id, re);
                continue;
            }

            try {
                if (!processor.filter(jo)) {
                    continue;
                }
            } catch (RuntimeException re) {
                log.error("( {} ) filter operation error - {}", _id, re);
                continue;
            }

            try {
                processor.process(jo);
            } catch (RuntimeException re) {
                log.error("( {} ) process error - {}", _id, re);
            }
        }
        log.info("( {} ) exited Dispatcher process loop", _id);
    }

}
